allennlp/training/callbacks/track_epoch.py

```diff
@@ -0,0 +1,32 @@
+from typing import Dict, Any, TYPE_CHECKING
+
+from allennlp.training.callbacks.callback import TrainerCallback
+
+if TYPE_CHECKING:
+    from allennlp.training.trainer import GradientDescentTrainer
+
+
+@TrainerCallback.register("track_epoch_callback")
+class TrackEpochCallback(TrainerCallback):
+    """
+    A callback that you can pass to the `GradientDescentTrainer` to access the current epoch number
+    in your model during training. This callback sets `model.epoch`, which can be read inside of
+    `model.forward()`. We set `model.epoch = epoch + 1` which now denotes the number of
+    completed epochs at a given training state.
```

Comment:
It's set at 0 `on_start`, and only changes at the end of each epoch. I just copied and pasted this one over, I haven't changed anything with it.

Reply:
I figured you didn't, but it seems wrong. If `epoch` is set to `0` for the first batch, then `epoch + 1` is `1`, which is not the number of completed epochs.

I don't think we can change the code, because people might rely on the old behavior. But we should change the docs to be accurate.

Reply:
It's not changed on the first batch though. Is that what you're saying? It's only changed at the end of each epoch.

Reply:
Yes, I think `model.epoch` will be `1` during the entirety of the first epoch, while the first epoch has not been completed yet. That goes counter the wording of the docs.

Reply:
`model.epoch` will be 0 for the first epoch. It's set `on_start` (https://github.com/allenai/allennlp/blob/2185ec016890c1f856440cfcf5f01a491da4a3dc/allennlp/training/callbacks/track_epoch.py#L18-L22), and is unchanged until after the first epoch is completed, at which point it will be incremented through the `on_epoch` method (https://github.com/allenai/allennlp/blob/2185ec016890c1f856440cfcf5f01a491da4a3dc/allennlp/training/callbacks/track_epoch.py#L24-L32).

Reply:
Ah, cool! Thanks for clearing that up.